### PR TITLE
New subspace configuration calculation.

### PR DESCRIPTION
**Pull request by weinbe58**
This new implementation starts from a single site and builds the subspace by adding sites one-by-one, updating the many-body fock states gradually instead of the original implementation, which started with the full Fock space and removed states that violate the blockade constraint. This means that the implementation is not only faster but also more memory efficient. 

The old implementation calculating the fock space for a PXP chain 30 sites benchmarks at:

`4688 function calls in 31.890 seconds`

vs. the new implementation takes 

`12861 function calls in 0.101 seconds`

 Naive implementation (could be ported to Julia)

```python


def draft_0(blockaded_sites, states):
    def is_rydberg(state, i):
        return bool((state >> i) & 1)

    configurations = list(states)

    for j, sites in enumerate(blockaded_sites, 1):
        new_configurations = []

        blockaded = [
            config
            for config in configurations
            if any(is_rydberg(config, i) for i in sites)
        ]
        non_blockaded = [
            config
            for config in configurations
            if not any(is_rydberg(config, i) for i in sites)
        ]

        for config in non_blockaded:
            for state in states:
                new_config = config + (state << j)
                new_configurations.append(new_config)

        for config in blockaded:
            for state in states[:-1]:
                new_config = config + (state << j)
                new_configurations.append(new_config)

        configurations = new_configurations

    configurations = sorted(configurations)

```

### CODE REVIEW

**Comment by Roger-luo** [comment]
oh, this is nice! This is more fine-grained now.

File: src/bloqade/emulate/ir/space.py [hunk 1]
```diff
@@ -33,44 +33,57 @@ def create(cls, register: "Register"):
...
+            for index_2, site_2 in enumerate(sites[:index_1]):
                 site_2 = np.asarray(site_2)
                 if np.linalg.norm(site_1 - site_2) <= blockade_radius:
                     atoms.append(index_2)
 
             check_atoms.append(atoms)
 
-        configurations = np.arange(Ns, dtype=np.min_scalar_type(Ns - 1))
+        min_int_type = np.min_scalar_type(Ns - 1)
+        config_type = np.result_type(min_int_type, np.uint32)
```
Comment (Roger-luo) [inline src/bloqade/emulate/ir/space.py:47]: IIUC, this is requiring the minimum type to be `uint32`?

File: src/bloqade/emulate/ir/space.py [hunk 2]
```diff
@@ -33,44 +33,57 @@ def create(cls, register: "Register"):
...
-            # get which configurations are in rydberg state for the current index.
-            rydberg_configs_1 = atom_type.is_rydberg_at(configurations, index_1)
-            for index_2 in indices:  # loop over neighbors within blockade radius
-                # get which configus have the neighbor with a rydberg excitation
-                rydberg_configs_2 = atom_type.is_rydberg_at(configurations, index_2)
-                # get which states do not violate constraint
-                mask = np.logical_not(
-                    np.logical_and(rydberg_configs_1, rydberg_configs_2)
+        states = np.arange(atom_type.n_level, dtype=config_type)
+        configurations = states.copy()
```
Comment (Roger-luo) [inline src/bloqade/emulate/ir/space.py:55]: why do we need to `copy` here?

File: src/bloqade/emulate/ir/space.py [hunk 3]
```diff
@@ -33,44 +33,57 @@ def create(cls, register: "Register"):
...
+                np.ones_like(non_blockaded), states * atom_type.n_level**index_1
+            )
+            # add all but the rydberg state because some at least one is blockading
+            new_blockaded = np.kron(blockaded, np.ones_like(states[:-1])) + np.kron(
+                np.ones_like(blockaded), states[:-1] * atom_type.n_level**index_1
+            )
+
+            configurations = np.hstack((new_blockaded, new_non_blockaded))
+
+        configurations.sort()
```
Comment (Roger-luo) [inline src/bloqade/emulate/ir/space.py:86]: it seems the only mutation happens here?

Comment (weinbe58) [reply src/bloqade/emulate/ir/space.py:86]: I'll double check this.

File: src/bloqade/emulate/ir/space.py (hunk 1 above)
Comment (Roger-luo) [reply src/bloqade/emulate/ir/space.py:47]: I think that's fine, just want to make sure I don't misunderstand the purpose.

**Review by Roger-luo — APPROVED** [review]

**Pull request merged**